### PR TITLE
Selected in View within Gallery

### LINKED ISSUE #4778: [enhancement]: Keep Selected in View within Gallery
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Contact Details

steinn on discord

### What should this feature add?

A way to easily know and find the selected image in the gallery. As you browse the gallery using ← & → arrow keys, as one can quickly lose track of where in the list the selected one is. Auto-scrolling to follow along the selected as it goes down or up, always keeping it in view, or some button to scroll to where it is.

I found myself wanting to drag multiple images in a row onto another board only to drag the wrong image(s) or else hard to find where I am / was.

Currently if you repeatedly click the right arrow to view previous images the selected border quickly moves down and out of view depending on screen resolution and zoom %.

Might be worth it to have it as a toggleable setting to have it always in view.

### Alternatives

one could have a little indicator on the scroll where like a circle or something next to it indicating where one would scroll to to find it. 

### Additional Content

_No response_

### PR DESCRIPTION
**Pull request by rohinish404**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [X] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission

## Have you discussed this change with the InvokeAI team?
- [X] Yes
- [ ] No, because:

      
## Have you updated all relevant documentation?
- [ ] Yes
- [ ] No

## Related Tickets & Documents

- Related Issue #4778 
- Closes #4778 

## Added/updated tests?

- [ ] Yes
- [ ] No : _please replace this line with details on why tests
      have not been included_

### CODE REVIEW

**Review by lstein — CHANGES_REQUESTED** [review]
Tested and it works as advertised. However, the behavior is that the selected thumbnail is always on the top row of the gallery. I would personally prefer behavior in which the gallery scrolls up to keep the selected thumbnail in view _only after the selected thumbnail reaches the bottom of the gallery_. I find it useful to see the selected thumbnail in context of what both precedes and follows it.

File: invokeai/frontend/web/src/features/gallery/components/ImageGrid/GalleryImageGrid.tsx [hunk 1]
```diff
@@ -48,6 +58,10 @@ const GalleryImageGrid = () => {
   const { currentViewTotal } = useBoardTotal(selectedBoardId);
   const queryArgs = useAppSelector(selectListImagesBaseQueryArgs);
 
+  const lastSingleSelectionImage = useAppSelector(selector);
+
+  const imgRef = useRef<VirtuosoGridHandle>(null);
```
Comment (psychedelicious) [inline invokeai/frontend/web/src/features/gallery/components/ImageGrid/GalleryImageGrid.tsx:63]: Suggest a `virtuosoRef` as a clearer name for this

**Comment by psychedelicious** [comment]
I agree with @lstein, it should only scroll when it needs to. I think you can use the `rangeChanged` callback to accomplish this by comparing the current visible range with the selected item index.

**Comment by rohinish404** [comment]
> Tested and it works as advertised. However, the behavior is that the selected thumbnail is always on the top row of the gallery. I would personally prefer behavior in which the gallery scrolls up to keep the selected thumbnail in view _only after the selected thumbnail reaches the bottom of the gallery_. I find it useful to see the selected thumbnail in context of what both precedes and follows it.

@lstein Fixed with latest commit!

File: invokeai/frontend/web/src/features/gallery/components/ImageGrid/GalleryImageGrid.tsx (hunk 1 above)
Comment (rohinish404) [reply invokeai/frontend/web/src/features/gallery/components/ImageGrid/GalleryImageGrid.tsx]: fixed with latest commit!

**Comment by psychedelicious** [comment]
Marking this as a draft pending resolution of issues as discussed in https://discord.com/channels/1020123559063990373/1177849030911479808

**Review by lstein — APPROVED** [review]
I've tested the recent changes and everything seemed to work:
1. Selecting next and previous thumbnails using right and left arrow keys.
2. Scrolling up and down in the gallery using up/down arrow keys
3. Scrolling with mouse wheel
4. Scrolling current selected thumbnail out of sight and then using right arrow to select next thumbnail scrolls the selected thumbnail into view.

The only oddity I noticed is that when you select a thumbnail it is outlined in blue as expected. However, when using the right arrow to select the next thumbnail, the outline turns to white. This is also present in `main`, so not specific to the current PR.

**Comment by rohinish404** [comment]
> I've tested the recent changes and everything seemed to work:
> 
> 1. Selecting next and previous thumbnails using right and left arrow keys.
> 2. Scrolling up and down in the gallery using up/down arrow keys
> 3. Scrolling with mouse wheel
> 4. Scrolling current selected thumbnail out of sight and then using right arrow to select next thumbnail scrolls the selected thumbnail into view.
> 
> The only oddity I noticed is that when you select a thumbnail it is outlined in blue as expected. However, when using the right arrow to select the next thumbnail, the outline turns to white. This is also present in `main`, so not specific to the current PR.

Thanks for the review!! So no further changes are required?

**Review by psychedelicious — APPROVED** [review]
I updated the branch and made two small improvements:
- In `useNextPrevImage`, do not create new variables for all the refs. We can use them directly to reduce memory usage and improve performance. It's a tiny improvement.
- In `useNextPrevImage`, the the prev / next callbacks only aligned to 'start' and 'end' respectively. This resulted in the alignment being different depending on which arrow you press, but that's not correct. I extracted the logic from `useScrollToVisible` that determines which alignment to use, and also used it in `useNextPrevImage`

I also make two changes related to network requests:
- Set the fallback debounce of metadata fetching hook of 300ms (was 0ms)
- Add a debounced hook for workflow, and set it to 300ms

This *drastically* reduces the number of network requests made when navigating the gallery with arrow keys without impacting user experience.

Thanks for working on this @rohinish404 !

**Comment by rohinish404** [comment]
> I updated the branch and made two small improvements:
> 
> * In `useNextPrevImage`, do not create new variables for all the refs. We can use them directly to reduce memory usage and improve performance. It's a tiny improvement.
> * In `useNextPrevImage`, the the prev / next callbacks only aligned to 'start' and 'end' respectively. This resulted in the alignment being different depending on which arrow you press, but that's not correct. I extracted the logic from `useScrollToVisible` that determines which alignment to use, and also used it in `useNextPrevImage`
> 
> I also make two changes related to network requests:
> 
> * Set the fallback debounce of metadata fetching hook of 300ms (was 0ms)
> * Add a debounced hook for workflow, and set it to 300ms
> 
> This _drastically_ reduces the number of network requests made when navigating the gallery with arrow keys without impacting user experience.
> 
> Thanks for working on this @rohinish404 !

Woah..that's some awesome improvements! Thanks for merge and helping out throughout this issue! @psychedelicious and @lstein

**Pull request merged**